File: frontend/src/utils/passwordValidator.ts
export interface PasswordRequirements {
  minLength: boolean;
  hasUpperCase: boolean;
  hasLowerCase: boolean;
  hasNumber: boolean;
  hasSpecialChar: boolean;
}

export const PASSWORD_RULES = {
  minLength: 8,
  requireUpperCase: true,
  requireLowerCase: true,
  requireNumber: true,
  requireSpecialChar: true,
};

export const validatePassword = (password: string): PasswordRequirements => {
  return {
    minLength: password.length >= PASSWORD_RULES.minLength,
    hasUpperCase: /[A-Z]/.test(password),
    hasLowerCase: /[a-z]/.test(password),
    hasNumber: /\d/.test(password),
    hasSpecialChar: /[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?]/.test(password),
  };
};

export const isPasswordValid = (password: string): boolean => {
  const requirements = validatePassword(password);
  return (
    requirements.minLength &&
    requirements.hasUpperCase &&
    requirements.hasLowerCase &&
    requirements.hasNumber &&
    requirements.hasSpecialChar
  );
};

export const getPasswordErrorMessage = (): string[] => {
  return [
    `At least ${PASSWORD_RULES.minLength} characters`,
    'At least one uppercase letter',
    'At least one lowercase letter',
    'At least one number',
    'At least one special character (!@#$%^&* etc.)',
  ];
};

File: backend/auth-service/src/Api/Auth/PasswordValidator.cs
using System.Text.RegularExpressions;

namespace Api.Auth
{
    public class PasswordValidator
    {
        private const int MinLength = 8;
        private const string UpperCasePattern = @"[A-Z]";
        private const string LowerCasePattern = @"[a-z]";
        private const string NumberPattern = @"\d";
        private const string SpecialCharPattern = @"[!@#$%^&*()_+=\[\]{};':\|,.<>\\/?-]";

        public static (bool IsValid, List<string> Errors) ValidatePassword(string password)
        {
            var errors = new List<string>();

            if (string.IsNullOrWhiteSpace(password) || password.Length < MinLength)
                errors.Add($"Password must be at least {MinLength} characters long");

            if (!Regex.IsMatch(password, UpperCasePattern))
                errors.Add("Password must contain at least one uppercase letter");

            if (!Regex.IsMatch(password, LowerCasePattern))
                errors.Add("Password must contain at least one lowercase letter");

            if (!Regex.IsMatch(password, NumberPattern))
                errors.Add("Password must contain at least one number");

            if (!Regex.IsMatch(password, SpecialCharPattern))
                errors.Add("Password must contain at least one special character (!@#$%^&* etc.)");

            return (errors.Count == 0, errors);
        }
    }
}
